### PR TITLE
add mkldnn softmax backward 

### LINKED ISSUE #13365: Missing MKLDNN implementation for backward softmax operator
## Description
While adding unit tests for various MKLDNN operators, I discovered that the backward softmax operator does not have an MKLDNN implementation. Is there any particular reason why it is not currently there? Is it missing and supposed to be added in the future?

## Environment info (Required)
https://github.com/apache/incubator-mxnet/blob/master/src/operator/nn/softmax.cc#L108
backward softmax operator does not have an associated (FComputeEx) MKLDNN implementation

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
add mkldnn softmax backward implementation
unitest pass

Should fix #13365 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

@PatricZhao @TaoLv

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_ops-inl.h [hunk 1]
```diff
@@ -93,6 +93,10 @@ void MKLDNNLeakyReluBackward(const nnvm::NodeAttrs &attrs, const OpContext &ctx,
 void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
                           const NDArray &in_data, const OpReqType &req,
                           const NDArray &out_data);
+void MKLDNNSoftmaxBackward(const nnvm::NodeAttrs& attrs, const OpContext &ctx,
+                          const std::vector<NDArray> &in_data,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_ops-inl.h]: indent.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 2]
```diff
@@ -86,6 +98,33 @@ void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
   stream->Submit();
 }
 
+void MKLDNNSoftmaxBackward(const nnvm::NodeAttrs& attrs,
+                          const OpContext &ctx,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: indent.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 3]
```diff
@@ -86,6 +98,33 @@ void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
...
+
+  auto out_mem = CreateMKLDNNMem(out_data[0], bwd_pd.diff_src_desc(), req[0]);
+  MKLDNNStream *stream = MKLDNNStream::Get();
+  mkldnn_args_map_t args = {
+    { MKLDNN_ARG_DST, *data_mem },
+    { MKLDNN_ARG_DIFF_DST, *diff_mem },
+    { MKLDNN_ARG_DIFF_SRC, *out_mem.second },
+  };
+
+  stream->RegisterPrimArgs(bwd_pd, args);
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: pd?

File: src/operator/nn/softmax.cc [hunk 4]
```diff
@@ -54,13 +54,30 @@ static void SoftmaxComputeExCPU(const nnvm::NodeAttrs& attrs,
                   inputs, req, outputs);
 }
 
+static void SoftmaxGradComputeExCPU(const nnvm::NodeAttrs& attrs,
+                                    const OpContext& ctx,
+                                    const std::vector<NDArray>& inputs,
+                                    const std::vector<OpReqType>& req,
+                                    const std::vector<NDArray>& outputs) {
+  // It seems MKLDNN softmax doesn't support training.
```
Comment (TaoLv) [inline src/operator/nn/softmax.cc]: Can you elaborate?

File: src/operator/nn/softmax.cc [hunk 5]
```diff
@@ -54,13 +54,30 @@ static void SoftmaxComputeExCPU(const nnvm::NodeAttrs& attrs,
...
+                  inputs, req, outputs);
+}
+
 inline static bool SoftmaxStorageType(const nnvm::NodeAttrs& attrs,
                                       const int dev_mask,
                                       DispatchMode* dispatch_mode,
                                       std::vector<int> *in_attrs,
                                       std::vector<int> *out_attrs) {
   const SoftmaxParam& param = nnvm::get<SoftmaxParam>(attrs.parsed);
-  CHECK_EQ(in_attrs->size(), (param.use_length.value()) ? 2U : 1U);
```
Comment (TaoLv) [inline src/operator/nn/softmax.cc]: Why remove this?

File: src/operator/nn/mkldnn/mkldnn_ops-inl.h (hunk 1 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_ops-inl.h]: Will change. Thanks

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 2 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: Will change. Thanks

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 3 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: will change it to cache primitive pd.

File: src/operator/nn/softmax.cc (hunk 4 above)
Comment (rongzha1) [reply src/operator/nn/softmax.cc]: Will remove this out-dated comments

File: src/operator/nn/softmax.cc (hunk 5 above)
Comment (rongzha1) [reply src/operator/nn/softmax.cc]: this check will result in backward check fail. 
Will recover this check and add another function for backward.

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 3 above)
Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: I mean here you need give a primitive not a primitive descriptor. Please check the definition of RegisterPrimArgs.

Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: yes, you're right.
Although it is a pd, it call implicit constructor
inline primitive::primitive(const primitive_desc &pd) : primitive(pd.get()) {}
to get a primitive

Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: I will make it more clearly by adding  constructor before used.

**Comment by pengzhao-intel** [comment]
@rongzha1 please retrigger the PR @TaoLv could you help review again?

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 6]
```diff
@@ -131,6 +143,80 @@ void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
...
+
+  const mkldnn::softmax_backward &GetBwd() const {
+    return *bwd_;
+  }
+
+ private:
+  std::shared_ptr<mkldnn::softmax_backward> bwd_;
+};
+
+typedef ParamOpSign<SoftmaxParam> MKLDNNSoftmaxSignature;
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: Same as L99?

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 7]
```diff
@@ -131,6 +143,80 @@ void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
...
+    MKLDNNSoftmaxBwd bwd(*diff_mem, *data_mem, real_axis, fwd_pd);
+    it = AddToCache(&bwds, key, bwd);
+  }
+  return it->second;
+}
+
+void MKLDNNSoftmaxBackward(const nnvm::NodeAttrs& attrs,
+                           const OpContext &ctx,
+                           const std::vector<NDArray> &in_data,
+                           const std::vector<OpReqType>& req,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]:
```suggestion
                           const std::vector<OpReqType> &req,
```

File: src/operator/nn/mkldnn/mkldnn_softmax.cc [hunk 8]
```diff
@@ -131,6 +143,80 @@ void MKLDNNSoftmaxForward(const nnvm::NodeAttrs& attrs,
...
+                           const std::vector<NDArray> &out_data) {
+  if (req[0] == kNullOp) return;
+  CHECK_EQ(in_data.size(), 2U);
+  const SoftmaxParam& param = nnvm::get<SoftmaxParam>(attrs.parsed);
+  int axis = CheckAxis(param.axis, in_data[1].shape().ndim());
+  auto diff_mem = in_data[0].GetMKLDNNData();
+  auto data_mem = in_data[1].GetMKLDNNData();
+  auto bwd = GetSoftmaxBwd(param, axis, in_data, out_data);
+
+  auto out_mem = CreateMKLDNNMem(out_data[0], bwd.pd.diff_src_desc(), req[0]);
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_softmax.cc]: 1. Please check if you want to support req=kAddTo;
2. softmax backward primitive should support in-place calculation so no need to create additional buffer.

Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: you are right, create mem is used to support kAddTo

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 6 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: exactly same, add extra one when rebase, will remove this line Thanks

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 8 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: yes, original softmax bwd support kAddTo : SoftmaxGradCompute-->SoftmaxGrad -->KERNEL_ASSIGN(igrad[base + j*sa], Req, final_result);

**Comment by pengzhao-intel** [comment]
@TaoLv were all concerns resolved now?

File: src/operator/nn/mkldnn/mkldnn_softmax.cc (hunk 7 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_softmax.cc]: done
